<?xml version="1.0" encoding="UTF-8"?>

<launch>
  <!-- Simulation parameters -->
  <arg name="use_sim_time" default="true" />
  <arg name="gui" default="true" />
  <arg name="headless" default="false" />
  <arg name="world_name" default="$(find dynamics_planner_nav)/launch/data/world_files/BARN/world_1.world" />

  <!-- Program selection: RunMP, SwitchMP, or None (no planner) -->
  <arg name="program" default="SwitchMP" />

  <!-- Planner selection (empty string = wait for /planner_switch topic) -->
  <arg name="planner" default="" />

  <!-- Robot configuration -->
  <arg name="front_laser" default="false" />
  <arg name="default_config" value="front_laser" if="$(arg front_laser)" />
  <arg name="default_config" value="vlp_laser" unless="$(arg front_laser)" />
  <arg name="config" default="$(arg default_config)" />

  <!-- Optional features -->
  <arg name="joystick" default="false" />
  <arg name="use_gmapping" default="false" />
  <arg name="use_rviz" default="true" />
  <arg name="rviz_config" default="$(find dynamics_planner_nav)/rviz/a.rviz" />

  <!-- Launch Gazebo with the specified world -->
  <include file="$(find gazebo_ros)/launch/empty_world.launch">
    <arg name="debug" value="0" />
    <arg name="gui" value="$(arg gui)" />
    <arg name="use_sim_time" value="$(arg use_sim_time)" />
    <arg name="headless" value="$(arg headless)" />
    <arg name="world_name" value="$(arg world_name)" />
  </include>

  <!-- Spawn Jackal -->
  <include file="$(find jackal_gazebo)/launch/spawn_jackal.launch">
    <arg name="x" value="-2" />
    <arg name="y" value="3" />
    <arg name="z" value="1.0" />
    <arg name="yaw" value="0" />
    <arg name="config" value="$(arg config)" />
    <arg name="joystick" value="$(arg joystick)" />
  </include>

  <!-- Collision check publisher (only if program is not None) -->
  <group unless="$(eval program == 'None')">
    <node pkg="dynamics_planner_nav" type="collision_check_publisher_node" name="collision_check_publisher_node" />
  </group>

  <!-- Launch planner: RunMP mode (requires planner to be specified) -->
  <group if="$(eval program == 'RunMP' and planner != '')">
    <node pkg="dynamics_planner_nav" type="dynamics_planner_nav_node" name="dynamics_planner_nav_node" output="screen"
      args="RunMP $(find dynamics_planner_nav)/data/ParamsForJackalGlobal.txt UseMP $(arg planner)"/>
  </group>

  <!-- Launch planner: SwitchMP mode (planner can be empty, waits for /planner_switch topic) -->
  <group if="$(eval program == 'SwitchMP')">
    <node pkg="dynamics_planner_nav" type="dynamics_planner_nav_switch_node" name="dynamics_planner_nav_switch_node" output="screen">
      <param name="planner" value="$(arg planner)" if="$(eval planner != '')" />
    </node>
  </group>

  <!-- RViz visualization -->
  <group if="$(arg use_rviz)">
    <param name="use_gui" value="true"/>
    <node name="rviz" pkg="rviz" type="rviz" args="-d $(arg rviz_config)" />
  </group>

  <!-- GMapping SLAM (optional) -->
  <group if="$(arg use_gmapping)">
    <arg name="scan_topic" default="$(eval optenv('JACKAL_LASER_TOPIC', 'front/scan'))" />
    <include file="$(find jackal_navigation)/launch/include/gmapping.launch">
      <arg name="scan_topic" value="$(arg scan_topic)" />
    </include>
  </group>

</launch>
